pallets/basic-sto/src/lib.rs

```diff
@@ -0,0 +1,178 @@
+        /// Sender does not have required permissions
+        Unauthorized,
+        /// An arithmetic operation overflowed
+        Overflow,
+        /// Not enough tokens left for sale
+        InsufficientTokensRemaining
+    }
+}
+
+decl_storage! {
+    trait Store for Module<T: Trait> as StoCapped {
+        /// All fundraisers that are currently running. (ticker, fundraiser_id) -> Fundraiser
+        Fundraisers get(fn fundraisers): double_map hasher(blake2_128_concat) Ticker, hasher(twox_64_concat) u64 => Fundraiser<T::Balance>;
+        /// Total fundraisers created for a token
+        FundraiserCount get(fn fundraiser_count): map hasher(twox_64_concat) Ticker => u64;
+    }
+}
+
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        type Error = Error<T>;
+
+        fn deposit_event() = default;
+
+        /// Create a new offering.
+        #[weight = 200_000_000]
+        pub fn create_fundraiser(
+            origin,
+            offering_token: Ticker,
+            raise_token: Ticker,
+            sell_amount: T::Balance,
+            price_per_token: u128,
+            venue_id: u64
+        ) -> DispatchResult {
+            let sender = ensure_signed(origin)?;
+            let did = Context::current_identity_or::<Identity<T>>(&sender)?;
+            ensure!(T::Asset::is_owner(&offering_token, did), Error::<T>::Unauthorized);
+            // TODO: Take custodial ownership of $sell_amount of $offering_token from treasury?
+            let fundraiser_id = Self::fundraiser_count(offering_token) + 1;
+            <Fundraisers<T>>::insert(
+                offering_token,
+                fundraiser_id,
+                Fundraiser {
+                    raise_token,
+                    price_per_token,
+                    venue_id,
+                    remaining_amount: sell_amount
+                }
+            );
+            Self::deposit_event(
+                RawEvent::FundraiserCreated(did, offering_token, raise_token, sell_amount, price_per_token, venue_id, fundraiser_id)
+            );
+            Ok(())
+        }
+
+        /// Create a new offering.
+        #[weight = 200_000_000]
+        pub fn invest(origin, offering_token: Ticker, fundraiser_id: u64, offering_token_amount: T::Balance) -> DispatchResult {
+            let sender = ensure_signed(origin.clone())?;
+            let did = Context::current_identity_or::<Identity<T>>(&sender)?;
+            let mut fundraiser = Self::fundraisers(offering_token, fundraiser_id);
+            ensure!(fundraiser.remaining_amount >= offering_token_amount, Error::<T>::InsufficientTokensRemaining);
+            // Ceil of offering_token_amount * price_per_million
+            let raise_token_amount = offering_token_amount
+                .checked_mul(&fundraiser.price_per_token.into())
+                .ok_or(Error::<T>::Overflow)?
+                .saturating_add(999_999.into())
+                / 1_000_000.into();
```

Comment:
Could use the constant POLYX or Dollar

Reply:
I think this is clearer since we have `999_999` above. I can do `Dollar - 1` but this is fine for now.

Reply:
`- 1` would definitely be clearer :)

Also, extract this to an utility function for `ceil`?

Reply:
I'll use the `ONE_UNIT` constant but hold off creating a utility function since we only use it once and the type implementation comes from upstream.